src/rgw/rgw_rest_s3.cc

```diff
@@ -3238,7 +3240,14 @@ int RGW_Auth_S3::authorize(RGWRados *store, struct req_state *s)
       if (algorithm != "AWS4-HMAC-SHA256") {
         return -EPERM;
       }
-      return authorize_v4(store, s);
+      /* compute first aws4 signature (stick to the boto2 implementation) */
+      int err = authorize_v4(store, s);
+      if (err && store->ctx()->_conf->rgw_s3_auth_aws4_presigned_url_bugfix) {
```

Comment:
ERR_SIGNATURE_NO_MATCH is used to return failure when signatures don't match. Bear in mind authorize_v4 calls more functions and more errors can be returned. For example, parsing errors, missed fields, etc. The approach here is checking the signature is ok (err = 0) vs we hit some error (err != 0). Returning the error var, we are propagating the result of the first authorize_v4 call. If err = 0 all things were okay with the first signature and the second signature computation didn't run.